homeassistant/components/statistics/sensor.py

```diff
@@ -393,13 +395,12 @@ async def async_start_preview(
         await self._async_stats_sensor_startup()
         return self._call_on_remove_callbacks
 
-    @callback
-    def _async_stats_sensor_state_listener(
+    def _async_handle_new_state(
```

Comment:
One very important thing to keep in mind is that when we get a `state_reported` event, we get a reference to the same `State` object as in the last `state_update` event, Home Assistant mutates the existing object by updating the `last_reported` timestamp and then fires the event.

This means it's error prone to store the `State` object. This integration does not, it instead copies the `state` and `last_reported` timestamp so there's no problem right now.

To help prevent accidental future changes, we may want to amend the docstring of `_add_state_to_queue` to mention we should not store a reference to the state.

Also, maybe add a test which would break if the implementation is changed to store references to `State` objects?
The problem which would be caused by such a change is that calculations based on the `ages` bugger would be wrong. Maybe it's enough to check `age_coverage_ratio` in the `test_sensor_source_with_force_update` test?

Reply:
I have added a test that is sensitive to the timestamps. Identical timestamps of subsequent values would also cause a division by zero in the time-based average functions.